### PR TITLE
add refresh token functionality

### PR DESCRIPTION
**Pull request by jsachs**
#### What's this PR do?
Add refresh token endpoints and behavior.

#### Related JIRA tickets:
[SER-186](https://jira.amida-tech.com/browse/SER-186)

#### How should this be manually tested?
First, make sure the test harness runs.

Then, you will want to mimic the client workflow. Log in normally. You should set the normal auth token to have a short expiration time. When the token expires, confirm that you can use the reset token endpoint `POST /token` to get a new, valid JWT.

This behavior will be handled by clients. If they receive a 401 based on an expired token, they should call the `/token` endpoint then resubmit the 401. @mhiner can confirm that this is the correct design or correct me if I am wrong.

There is also an endpoint `POST /token/reject` that will expire your current refresh token. The JWT token will still be valid, but the only way to get a new refresh token is to log in again. This can be used by a client as part of a logout process if they wish.

#### Any background context you want to provide?
https://solidgeargroup.com/refresh-token-with-jwt-authentication-node-js

### CODE REVIEW

**Comment by sirctseb** [comment]
Should the refresh token be in a UserRefreshToken table? One column in the user's table will prevent a user from being refreshably logged in in two places.

File: config/passport.js [hunk 1]
```diff
@@ -24,31 +24,35 @@ const opts = {
...
 module.exports = (passport) => {
-    passport.use(new JwtStrategy(opts, (jwtPayload, done) => {
+    const jwtStrategy = new JwtStrategy(opts, (jwtPayload, done) => {
         User.findOne({ where: { username: jwtPayload.username } })
             .then(user => done(null, user))
             .catch(err => done(err, false));
-    }));
+    });
+
+    const fbStrategy = new FacebookStrategy({
```
Comment (sirctseb) [inline config/passport.js]: The FacebookStrategy constructor fails if the environment variables are not set, preventing the service from starting. If facebook config is optional, we should move the constructor back into the `if`

File: server/controllers/auth.controller.js [hunk 2]
```diff
@@ -44,17 +45,86 @@ function login(req, res, next) {
...
+        }
+        const userInfo = {
+            id: userResult.id,
+            username: userResult.username,
+            email: userResult.email,
+            scopes: userResult.scopes,
+        };
+
+        const jwtToken = signJWT(userInfo);
+        const refreshToken = randtoken.uid(128);
```
Comment (sirctseb) [inline server/controllers/auth.controller.js]: It's almost no extra work for the client, but is there a reason to generate a new refresh token here in addition to the access token? I haven't seen that done in my reading.

**Comment by jsachs** [comment]
@sirctseb it's not a bad idea for normalization, but don't we want to maintain a single auth service for each application? What's the scenario where a single auth service needs to allow a user to be logged in to two places at once?

File: server/controllers/auth.controller.js (hunk 2 above)
Comment (jsachs) [reply server/controllers/auth.controller.js]: I think the idea is that a refresh token should be single-use.

**Comment by sirctseb** [comment]
The link in the background does not address the topic of regenerating the refresh token itself on refresh, but by way of explaining lifetimes, it offers this:

> In order to cut a scenario like this one, a system must be implemented in the server to invalidate a refresh token, besides setting a lifetime that obviously must be longer than that of the access tokens.

The refresh token is described as long-lived but revokable.

In its implementation of the `/token` endpoint, it responds with only a new access token: `res.json({token: 'JWT ' + token})`

Here is auth0's high-level description of the refresh token (from https://auth0.com/learn/refresh-tokens/):

> A Refresh Token is a special kind of token that can be used to obtain a renewed access token —that allows accessing a protected resource— at any time. You can request new access tokens until the refresh token is blacklisted. Refresh tokens must be stored securely by an application because they essentially allow a user to remain authenticated forever.

The article goes on to say that due to security concerns, "it is not safe to store them in the browser." They do not elaborate and I don't know what they consider "in the browser." If that means no cookies, and the token is only kept in memory, then the refresh token will not survive a tab being closed, which doesn't sound right to me.

This more technical article (https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/) also does not offer a solution for client-side storage but there are suggestions in the comments to store in an https-only cookie with a restricted path. This is a client-side concern, however, and not and issue for this PR

File: server/controllers/auth.controller.js [hunk 3]
```diff
@@ -44,17 +46,80 @@ function login(req, res, next) {
...
+ * @param req
+ * @param res
+ * @param next
+ * @returns {*}
+ */
+function rejectRefreshToken(req, res, next) {
+    const params = _.pick(req.body, 'refreshToken');
+    RefreshToken.findOne({ where: { token: params.refreshToken } })
+    .then((tokenResult) => {
+        console.log(tokenResult);
```
Comment (sirctseb) [inline server/controllers/auth.controller.js]: Just the console logs to remove

**Pull request merged**